src/smclient/smclient.cpp

```diff
@@ -0,0 +1,599 @@
+/*
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <Poco/Pipe.h>
+#include <Poco/PipeStream.h>
+#include <Poco/Process.h>
+#include <Poco/StreamCopier.h>
+
+#include <utils/exception.hpp>
+#include <utils/grpchelper.hpp>
+
+#include "logger/logmodule.hpp"
+#include "smclient.hpp"
+#include "utils/convert.hpp"
+
+namespace aos::sm::client {
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+Error SMClient::Init(const config::Config& config, iam::provisionmanager::ProvisionManagerItf& provisionManager,
+    crypto::CertLoaderItf& certLoader, crypto::x509::ProviderItf& cryptoProvider,
+    iam::nodeinfoprovider::NodeInfoProviderItf& nodeInfoProvider,
+    sm::resourcemanager::ResourceManagerItf& resourceManager, sm::networkmanager::NetworkManagerItf& networkManager,
+    sm::logprovider::LogProviderItf& logProvider, monitoring::ResourceMonitorItf& resourceMonitor,
+    sm::launcher::LauncherItf& launcher, bool provisioningMode)
+{
+    LOG_DBG() << "Init SMClient";
+
+    mConfig           = config;
+    mNodeInfoProvider = &nodeInfoProvider;
+    mCertLoader       = &certLoader;
+    mCryptoProvider   = &cryptoProvider;
+    mProvisionManager = &provisionManager;
+    mResourceManager  = &resourceManager;
+    mNetworkManager   = &networkManager;
+    mLogProvider      = &logProvider;
+    mResourceMonitor  = &resourceMonitor;
+    mLauncher         = &launcher;
+    mProvisioningMode = provisioningMode;
+
+    return ErrorEnum::eNone;
+}
+
+Error SMClient::Start()
+{
+    LOG_DBG() << "Start SMClient";
+
+    std::lock_guard lock {mMutex};
```

Comment:
Always put guard lock at the beginning of function in case the whole function needs to be guarded.